### PR TITLE
[MF] [JH] Estimate Next Purchase

### LINKED ISSUE #11: 11. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items
## Summary
In order to advise users about when to purchase things, the app needs to be able to calculate that guess and store it a future   date.

The `calculateEstimate` function from the `@the-collab-lab/shopping-list-utils` module will help us know how many days in the future a purchase should happen. In order to use `calculateEstimate`, we need to know how many days have passed since the time the item was last purchased.

## Acceptance criteria
- [x] When the user purchases an item, the item’s `dateNextPurchased` property is calculated using the `calculateEstimate` function and saved to the Firestore database
  - [x] `dateNextPurchased` is saved as **a date**, not a number
- [x] A `getDaysBetweenDates` function is exported from `utils/dates.js`  and imported into `api/firebase.js`
  - [x] This function takes two JavaScript Dates and return the number of days that have passed between them

### Working with Firestore’s dates
**Firestore returns dates as special objects called `Timestamps.`** You will need to convert the Timestamp to a JavaScript Date in order to work with it and complete this feature. The `Timestamp` object has a `.toDate()` method – see [the Firestore docs](https://firebase.google.com/docs/reference/js/v8/firebase.firestore.Timestamp#todate).

### Calculating the time between JavaScript dates
Comparing and manipulating JavaScript Dates is notoriously tricky. Here's a hint: all Date objects have a `getTime()` method that returns a representation of that date in milliseconds ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)).

### Calculating the next purchase date
The `@the-collab-lab/shopping-list-utils` module should already be installed in your project. This module exports `calculateEstimate`.

`calculateEstimate` accepts three arguments. They are, in order:
- `previousEstimate` (a number): The last estimated purchase interval.
- `daysSinceLastPurchase` (a number): The number of days that have elapsed between the last purchased date and now. You will need to figure out how to calculate this argument!
- `totalPurchases` (a number): The total number of purchases made for the item,

Given this information, `calculateEstimate` returns **a whole number** indicating the number of days until the user is likely to make their next purchase.

### PR DESCRIPTION
**Pull request by hsiangj**
## Description

This code updates the item's `dateNextPurchased` property in the `updateItem` API call by using the `calculateEstimate` function and a combination of the utils functions. The updated `dateNextPurchased` will be different for every item depending on the item's purchase history. The new date won't necessarily be at 7 or 14 or 30 day interval from the 'dateLastPurchased' as the new date is a result of `calculateEstimate`'s weighted estimate.

## Related Issue

Closes #11 

## Acceptance Criteria

- [x] When the user purchases an item, the item’s `dateNextPurchased` property is calculated using the `calculateEstimate` function and saved to the Firestore database
  - [x] `dateNextPurchased` is saved as **a date**, not a number
- [x] A `getDaysBetweenDates` function is exported from `utils/dates.js`  and imported into `api/firebase.js`
  - [x] This function takes two JavaScript Dates and return the number of days that have passed between them

## Type of Changes

Use one or more labels to help your team understand the nature of the change(s) you’re proposing. E.g., `bug fix` or `enhancement` are common ones.

## Updates
No UI feature.
## Testing Steps / QA Criteria

General testing to observe update in `dateNextPurchased`:
1. To test that this feature is working correctly, go to Firestore. Navigate to a specific item in a collection/list. Take note of the current `dateNextPurchased`.
2. Select the same list from the Home page in the app.
3. Navigate to the List page.
4. Check off the chosen item. 
5. `dateNextPurchased` should be updated to a new date in Firestore.

(Previously step 6) Further testing to observe different date ranges in `dateNextPurchased`: 
- Newly added items from today or yesterday with less than 2 purchases: `dateNextPurchased` should be updated to today's date when purchased.
- To observe different result in `dateNextPurchased` beyond today's date: manipulate `dateCreated` before checking off the item OR check off an item that was added to the list a while ago. `dateNextPurchased` should be updated to some date in the future.

### CODE REVIEW

**Comment by amalyam** [comment]
In testing with @piecanoe we found that this feature was not working consistently. For example, sometimes after changing dateLastPurchased by clicking the checkbox on a new item in the app, the dateNextPurchased in Firestore would update to be the same date. We found that changing dateCreated, as suggested in testing step #6, had no impact on dateNextPurchased, nor did changing dateLastPurchased directly in Firestore. We did find that sometimes dateNextPurchased seemed to update appropriately after checking the box for an item in the app, but we were not sure what caused it to work correctly.

**Comment by krsnamara** [comment]
Im not sure but I think what is happening is if you create a new item and then chk it as 'purchased' it has a very short time window to estimate when you will need the item again since the time between creation and purchase is so short. So when manipulating the dateCreated to something a few days prior or a month prior, then checking the item as purchased it will give a more accurate estimate of when to purchase next. Not sure how to change the configuration of the "algorithm" to produce a more accurate estimate with such little data to feed the machine. The items that had been previously created more than a few days prior should illicit the more accurate estimate response (which is recorded as dateNextPurchased)

**Comment by hsiangj** [comment]
Thanks for surfacing this! Similar to what Marty said, getting the same date for a new/recent item is expected based on our interpretation of what's happening under the hood. Just added additional details in the testing steps to help clarify! 
As for "... found that changing dateCreated, as suggested in testing step 6, had no impact on dateNextPurchased," do you mean that after changing the dateCreated and checking off the item in the app, the dateNextPurchased did not change?

**Comment by amalyam** [comment]
Thanks for updating the testing steps to make them more specific. App currently working as laid out there. @piecanoe found that there still seems to be some unusual behavior in the way the dates are updating (ex: an item marked needed "not too soon" will update to the current/today's date after being checked), but this seems to be outside the scope of this issue. Nice work despite the challenges!

**Review by amalyam — APPROVED** [review]
Thanks for updating the testing steps to make them more specific. App currently working as laid out there. @piecanoe found that there still seems to be some unusual behavior in the way the dates are updating (ex: an item marked needed "not too soon" will update to the current/today's date after being checked), but this seems to be outside the scope of this issue. Nice work despite the challenges!

File: src/api/firebase.js [hunk 1]
```diff
@@ -213,12 +213,42 @@ export async function addItem(listPath, { itemName, daysUntilNextPurchase }) {
...
+export async function updateItem(
+	listPath,
+	itemId,
+	dateLastPurchased,
+	totalPurchases,
+	dateNextPurchased,
+	dateCreated,
+) {
+	const dateNextPurchasedJS = dateNextPurchased.toDate();
+	const dateLastPurchasedJS = dateLastPurchased?.toDate();
```
Comment (rmorabia) [inline src/api/firebase.js:264]: Quick questions here: Why are these names with JS at the end? Also, what are the consequences of this line in particular being nullable?

**Review by rmorabia — APPROVED** [review]
Code looks good to me, but I will test this in a few hours when I can reach Firestore.

File: src/api/firebase.js (hunk 1 above)
Comment (hsiangj) [reply src/api/firebase.js:264]: We added JS to the end of the names to help identify Firestore timestamps that have been converted to JavaScript dates with the toDate() function, which were later used to update or calculate other time related variables. 
As for line 225, we realized that new items added to the list have a 'null' value in the 'dateLastPurchased' field, and without using optional chaining, the component threw an error.

**Pull request merged**